app/models/user.rb

```diff
@@ -84,7 +84,30 @@ def self.find_for_lti(auth_hash, signed_in_resource=nil)
       Course.create :context_id => class_id, :label => auth_hash.extra.consumer.context_label, :title => class_name unless class_name.nil?
     end
 
-    self.find_or_create_by_username_or_email(auth_hash.uid, auth_hash.info.email)
+    self.find_or_create_by_username_or_email(auth_hash.info.email, auth_hash.info.email)
+  end
+
+  def self.find_for_api(username, email, signed_in_resource=nil)
+    User.find_by_username(username) ||
+    User.find_by_email(email) ||
+    User.create(:username => username, :email => email)
+  end
+
+  def self.find_for_shibboleth(auth_hash, signed_in_resource=nil)
```

Comment:
Is this the best approach? For Shibboleth authentication, this is inspired from a21578b, b6cd19d & 30d83fc whereby the e-mail field is used for lookup in the `user` table an if doesn't exist then a new entry is created with the `username` as the e-mail address. This fails if the e-mail field is not part of the Shibboleth idp response.

Is this the best approach?